File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/recipeManipulation/adapter/viewHolder/RecipeManipulationPreparationStepItemViewHolder.java
package de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.adapter.viewHolder;

import android.support.v7.widget.RecyclerView;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.karzek.diettracker.R;
import de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.adapter.itemWrapper.RecipeManipulationViewItemWrapper;

/**
 * Created by MarjanaKarzek on 30.05.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 30.05.2018
 */
public class RecipeManipulationPreparationStepItemViewHolder extends RecyclerView.ViewHolder {

    @BindView(R.id.preparation_step_number) TextView stepNumber;
    @BindView(R.id.preparation_step_text) EditText description;

    private final OnDeletePreparationStepClickedListener onDeletePreparationStepClickedListener;
    private final OnEditPreparationStepFinishedListener onEditPreparationStepFinishedListener;

    public RecipeManipulationPreparationStepItemViewHolder(ViewGroup viewGroup,
                                                           OnDeletePreparationStepClickedListener onDeletePreparationStepClickedListener,
                                                           OnEditPreparationStepFinishedListener onEditPreparationStepFinishedListener) {
        super(LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.viewholder_recipe_man_preparation_step, viewGroup, false));
        ButterKnife.bind(this, itemView);

        this.onDeletePreparationStepClickedListener = onDeletePreparationStepClickedListener;
        this.onEditPreparationStepFinishedListener = onEditPreparationStepFinishedListener;

        setupDescriptionChangedListener();
    }

    private void setupDescriptionChangedListener() {
        description.setOnEditorActionListener(new EditText.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    onEditPreparationStepFinishedListener.onEditPreparationStepFinished((int) itemView.getTag(), description.getText().toString());
                }
                return true;
            }
        });
        description.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                onEditPreparationStepFinishedListener.onEditPreparationStepFinished((int) itemView.getTag(), description.getText().toString());
            }
        });
    }

    public void bind(RecipeManipulationViewItemWrapper item, int innerListId) {
        stepNumber.setText("" + item.getPreparationStepDisplayModel().getStepNo());
        description.setText(item.getPreparationStepDisplayModel().getDescription());

        itemView.setTag(innerListId);
    }

    @OnClick(R.id.action_delete) public void onDeletePreparationStepClicked() {
        onDeletePreparationStepClickedListener.onDeletePreparationStepClicked((int) itemView.getTag());
    }

    public interface OnDeletePreparationStepClickedListener {
        void onDeletePreparationStepClicked(int id);
    }

    public interface OnEditPreparationStepFinishedListener {
        void onEditPreparationStepFinished(int id, String description);
    }

}

File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/recipeManipulation/adapter/viewHolder/RecipeManipulationPreparationStepItemViewHolder.kt
package de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.adapter.viewHolder

import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.recyclerview.widget.RecyclerView
import de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.adapter.itemWrapper.RecipeManipulationViewItemWrapper
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.viewholder_recipe_man_preparation_step.*

/**
 * Created by MarjanaKarzek on 30.05.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 30.05.2018
 */
class RecipeManipulationPreparationStepItemViewHolder(override val containerView: View,
                                                      private val onDeletePreparationStepClickedListener: OnDeletePreparationStepClickedListener,
                                                      private val onEditPreparationStepFinishedListener: OnEditPreparationStepFinishedListener) : RecyclerView.ViewHolder(containerView), LayoutContainer {

    private fun setupDescriptionChangedListener() {
        preparation_step_text.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                onEditPreparationStepFinishedListener.onEditPreparationStepFinished(itemView.tag as Int, preparation_step_text.text.toString())
            }
            true
        }
        preparation_step_text.onFocusChangeListener = View.OnFocusChangeListener { _, _ -> onEditPreparationStepFinishedListener.onEditPreparationStepFinished(itemView.tag as Int, preparation_step_text.text.toString()) }
    }

    fun bind(item: RecipeManipulationViewItemWrapper, innerListId: Int) {
        preparation_step_number.text = "${item.preparationStepDisplayModel.stepNo}"
        preparation_step_text.setText(item.preparationStepDisplayModel.description)

        itemView.tag = innerListId

        setupDescriptionChangedListener()
        setupOnClickListeners()
    }

    private fun setupOnClickListeners() {
        action_delete.setOnClickListener { onDeletePreparationStepClickedListener.onDeletePreparationStepClicked(itemView.tag as Int) }
    }

    interface OnDeletePreparationStepClickedListener {
        fun onDeletePreparationStepClicked(id: Int)
    }

    interface OnEditPreparationStepFinishedListener {
        fun onEditPreparationStepFinished(id: Int, description: String)
    }

}
